client/ratings_reviews/components/ReviewFormModal.jsx

```diff
@@ -0,0 +1,147 @@
+import React from 'react';
+import $ from 'jquery';
+import ReviewFormCharacterisics from './ReviewFormCharacteristics.jsx';
+import helpers from '../helpers.js';
+
+class ReviewFormModal extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      rating: '',
+      summary: '',
+      body: '',
+      recommend: true,
+      name: '',
+      email: '',
+      photos: [],
+      characteristics: {}
+    };
+    this.handleChange = this.handleChange.bind(this);
+    this.handleSubmit = this.handleSubmit.bind(this);
+  }
+
+  handleChange(e) {
+    let name = e.target.name;
+    let value = e.target.value;
+    if (name.slice(0, 15) === 'characteristics') {
+      let currentCharacteristics = this.state.characteristics;
+      currentCharacteristics[e.target.className.toString()] = Number.parseInt(value);
+      value = currentCharacteristics;
+      name = name.slice(0, 15);
+    }
+    this.setState({
+      [name]: value
+    });
+  }
+
+  handleSubmit(e) {
+    e.preventDefault();
+    let data = {
+      'product_id': Number.parseInt(this.props.meta.product_id),
+      rating: Number.parseInt(this.state.rating),
+      summary: this.state.summary,
+      body: this.state.body,
+      recommend: JSON.parse(this.state.recommend),
+      name: this.state.name,
+      email: this.state.email,
+      photos: this.state.photos,
+      characteristics: this.state.characteristics
+    };
+    $.ajax({
+      url: '/reviews',
+      method: 'POST',
+      contentType: 'application/json; charset=utf-8',
+      data: JSON.stringify(data)
+    }).then(() => {
+      this.setState({
+        rating: '',
+        summary: '',
+        body: '',
+        recommend: true,
+        name: '',
+        email: '',
+        photos: [],
+        characteristics: {}
+      });
+      this.props.closeModal();
```

Comment:
Since setState can act asynchronously depending on how many things it's changing, not sure if you want closeModal to trigger only after it's done updating the state - if so, you could pass it as a callback to setState.

Reply:
Great point, I just changed it!